### ISSUE TITLE
Segmentation fault while training Logistic Regression

### ISSUE THREAD

**Issue by manuelmarucco** [VowpalWabbit/vowpal_wabbit#3833]
### Describe the bug

VW crashes with `Segmentation Fault 11` while training a logistic regression. We train a model with many different data sets and this error occurs sporadically. But if a dataset triggers the error it does so consistently. The attached file is one of the training sets which can be used to reproduce the error

The weirdest thing is that if you open the `train.txt` file and add or remove a character **anywhere** (after "|Text ") the training ends successfully but if you leave the file as it is or edit any character then it fails with segmentation fault. Attached is the compressed version of the `train.txt` file to reproduce the problem. The other thing to notice is that the `example counter` goes beyond the total number of examples in the file and that's when it's like it "diverges".

This happens with all versions of VW starting from 8.10.0 (even with 9.0.0) and in both macOS and Linux.
[train.txt.zip](https://github.com/VowpalWabbit/vowpal_wabbit/files/8371579/train.txt.zip)

### How to reproduce

To reproduce the problem you need the `train.txt` file which is simply a set of samples with all "x" that masks real data. The problem occurs with both real and fake data.

To train the model run the following command:

**Command to reproduce**
`$vw -d train.txt -f model.vw --loss_function=logistic -l 0.25 --l1 2e-05 --invert_hash feat.txt`

**Results**
using l1 regularization = 2e-05
final_regressor = model.vw
Num weight bits = 18
learning rate = 0.25
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = train.txt
num sources = 1
Enabled reductions: gd, scorer
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
0.902862 0.902862          353            1.6   1.0000  -0.7953        6
0.933825 0.959828          885            3.5   1.0000  -1.1263       10
0.817603 0.707740         1889            7.2   1.0000  -0.1800        4
0.810716 0.803828         3778           14.4  -1.0000   0.5680       13
0.766077 0.721437         7556           28.9  -1.0000   0.0585        7
0.743930 0.722397        16766           58.5   1.0000  -0.1409        5
0.724798 0.705666        33532          117.1  -1.0000   0.0396        3
0.708180 0.691562        69992          234.2  -1.0000   0.4146        6
0.703884 0.699587       137643          468.3  -1.0000   0.5214        8
0.698393 0.692902       279384          936.6  -1.0000   0.0094        5
0.695600 0.691693       471969         1874.0  unknown   0.1038        1
0.695600     n.a.       473844         3749.0  unknown   0.1038        1
0.695600     n.a.       477594         7499.0  unknown   0.1038        1
0.695600     n.a.       485094        14999.0  unknown   0.1038        1
0.695600     n.a.       500094        29999.0  unknown   0.1038        1
0.695600     n.a.       530094        59999.0  unknown   0.1038        1
0.695600     n.a.       590094       119999.0  unknown   0.1038        1
0.695600     n.a.       710094       239999.0  unknown   0.1038        1
0.695600     n.a.       950094       479999.0  unknown   0.1038        1
Segmentation fault: 11

### Version

8.10.0

### OS

macOS

### Language

CLI

**Comment by jackgerrits** [comment]
Looking into it.
